File: java/mapper/mapper2/src/com/ibatis/sqlmap/engine/cache/CacheModel.java
package com.ibatis.sqlmap.engine.cache;

import com.ibatis.common.resources.Resources;
import com.ibatis.sqlmap.engine.mapping.statement.ExecuteListener;
import com.ibatis.sqlmap.engine.mapping.statement.MappedStatement;
import net.sf.hibernate.exception.NestableRuntimeException;

import java.io.*;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Properties;
import java.util.Set;

public class CacheModel implements ExecuteListener {

  //private static final String CTX_LOCKED_OBJECTS = "__CACHE_MODEL_LOCKED_OBJECTS";

  private Object STATS_LOCK = new Object();
  private int requests = 0;
  private int hits = 0;

  /**
   * Constant to turn off periodic cache flushes
   */
  private static final long NO_FLUSH_INTERVAL = -99999;

  private String id;

  private boolean readOnly;
  private boolean serialize;

  private long lastFlush;
  private long flushInterval;
  private long flushIntervalSeconds;
  private Set flushTriggerStatements;

  private CacheController controller;

  private String resource;

  public CacheModel() {
    this.flushInterval = NO_FLUSH_INTERVAL;
    this.flushIntervalSeconds = NO_FLUSH_INTERVAL;
    this.lastFlush = System.currentTimeMillis();
    this.flushTriggerStatements = new HashSet();
  }

  public String getId() {
    return id;
  }

  public void setId(String id) {
    this.id = id;
  }

  public boolean isReadOnly() {
    return readOnly;
  }

  public void setReadOnly(boolean readOnly) {
    this.readOnly = readOnly;
  }

  public boolean isSerialize() {
    return serialize;
  }

  public void setSerialize(boolean serialize) {
    this.serialize = serialize;
  }

  public String getResource() {
    return resource;
  }

  public void setResource(String resource) {
    this.resource = resource;
  }

  public void setControllerClassName(String implementation)
      throws ClassNotFoundException, InstantiationException, IllegalAccessException {
    Class clazz = Resources.classForName(implementation);
    controller = (CacheController) Resources.instantiate(clazz);
  }

  /**
   * Getter for flushInterval property
   *
   * @return The flushInterval (in milliseconds)
   */
  public long getFlushInterval() {
    return flushInterval;
  }

  /**
   * Getter for flushInterval property
   *
   * @return The flushInterval (in milliseconds)
   */
  public long getFlushIntervalSeconds() {
    return flushIntervalSeconds;
  }

  /**
   * Setter for flushInterval property
   *
   * @param flushInterval The new flushInterval (in milliseconds)
   */
  public void setFlushInterval(long flushInterval) {
    this.flushInterval = flushInterval;
    this.flushIntervalSeconds = flushInterval / 1000;
  }

  /**
   * Adds a flushTriggerStatment. When a flushTriggerStatment is executed, the
   * cache is flushed (cleared).
   *
   * @param statementName The statement to add.
   */
  public void addFlushTriggerStatement(String statementName) {
    flushTriggerStatements.add(statementName);
  }

  /**
   * Gets an Iterator containing all flushTriggerStatment objects for this cache.
   *
   * @return The Iterator
   */
  public Iterator getFlushTriggerStatementNames() {
    return flushTriggerStatements.iterator();
  }

  /**
   * ExecuteListener event.  This will be called by a MappedStatement
   * for which this cache is registered as a ExecuteListener.  It will
   * be called each time an executeXXXXXX method is called.  In the
   * case of the Cache class, it is registered in order to flush the
   * cache whenever a certain statement is executed.
   * (i.e. the flushOnExecute cache policy)
   *
   * @param statement The statement to execute
   */
  public void onExecuteStatement(MappedStatement statement) {
    flush();
  }

  public double getHitRatio() {
    return (double) hits / (double) requests;
  }

  /**
   * Configures the cache
   *
   * @param props
   */
  public void configure(Properties props) {
    controller.configure(props);
  }

  /**
   * Clears the cache
   */
  public void flush() {
    lastFlush = System.currentTimeMillis();
    controller.flush(this);
  }

  /**
   * Get an object out of the cache.
   * A side effect of this method is that is may clear the cache if it has not been
   * cleared in the flushInterval.
   *
   * @param key The key of the object to be returned
   * @return The cached object (or null)
   */
  public Object getObject(CacheKey key) {
    synchronized (this) {
      if (flushInterval != NO_FLUSH_INTERVAL
          && System.currentTimeMillis() - lastFlush > flushInterval) {
        flush();
      }
    }

    Object value = controller.getObject(this, key);

    if (serialize && !readOnly && value != null) {
      try {
        ByteArrayInputStream bis = new ByteArrayInputStream((byte[]) value);
        ObjectInputStream ois = new ObjectInputStream(bis);
        value = ois.readObject();
        ois.close();
      } catch (Exception e) {
        throw new NestableRuntimeException("Error caching serializable object.  Cause: " + e, e);
      }
    }


    synchronized (STATS_LOCK) {
      requests++;
      if (value != null) {
        hits++;
      }
    }

    return value;
  }

  /**
   * Add an object to the cache
   *
   * @param key   The key of the object to be cached
   * @param value The object to be cached
   */
  public void putObject(CacheKey key, Object value) {
    if (serialize && !readOnly && value != null) {
      try {
        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        ObjectOutputStream oos = new ObjectOutputStream(bos);
        oos.writeObject((Serializable) value);
        oos.flush();
        oos.close();
        value = bos.toByteArray();
      } catch (IOException e) {
        throw new NestableRuntimeException("Error caching serializable object.  Cause: " + e, e);
      }
    }
    controller.putObject(this, key, value);
  }

}

File: cs/V2/mapper/Apache.Ibatis.DataMapper/Model/Cache/CacheModel.cs

#region Apache Notice
/*****************************************************************************
 * $Revision: 575913 $
 * $LastChangedDate$
 * $LastChangedBy$
 * 
 * iBATIS.NET Data Mapper
 * Copyright (C) 2006/2005 - The Apache Software Foundation
 *  
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * 
 ********************************************************************************/
#endregion

#region Using

using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Runtime.Serialization.Formatters.Binary;
using Apache.Ibatis.Common.Contracts;
using Apache.Ibatis.Common.Exceptions;
using Apache.Ibatis.Common.Logging;
using Apache.Ibatis.Common.Utilities;
using Apache.Ibatis.DataMapper.MappedStatements;

#endregion

namespace Apache.Ibatis.DataMapper.Model.Cache
{

    /// <summary>
    /// Summary description for CacheModel.
    /// </summary>
	[Serializable]
    [DebuggerDisplay("CacheModel: {Id}-{Implementation}")]
	public class CacheModel
	{
		#region Fields

		private readonly static IDictionary lockMap = new HybridDictionary();

		private static readonly ILog logger = LogManager.GetLogger( MethodBase.GetCurrentMethod().DeclaringType );
		/// <summary>
		/// This is used to represent null objects that are returned from the cache so 
		/// that they can be cached, too.
		/// </summary>
		public readonly static object NULL_OBJECT = new Object(); 

		/// <summary>
		/// Constant to turn off periodic cache flushes
		/// </summary>
		public const long NO_FLUSH_INTERVAL = -99999;

        private readonly object syncLock = new Object();
		private int requests = 0;
		private int hits = 0;
        private readonly string id = string.Empty;
		private ICacheController controller = null;
		private FlushInterval flushInterval = null;
		private long lastFlush = 0;
        private readonly string implementation = string.Empty;
		private readonly bool isReadOnly = true;
        private readonly bool isSerializable = false;
        private readonly IList<string> statementFlushNames = new List<string>();

		#endregion

		#region Properties
		/// <summary>
		/// Identifier used to identify the CacheModel amongst the others.
		/// </summary>
		public string Id
		{
			get { return id; }
		}


        /// <summary>
        /// Gets the statement flush on execute names.
        /// </summary>
        /// <value>The statement flush names.</value>
        public IList<string> StatementFlushNames
        {
            get { return statementFlushNames; }
        }

		/// <summary>
		/// Cache controller implementation name.
		/// </summary>
		public string Implementation
		{
			get { return implementation; }
		}

		/// <summary>
		/// Set the cache controller
		/// </summary>
		public ICacheController CacheController
		{
			set{ controller =value; }	
		}

		/// <summary>
		/// Set or get the flushInterval (in Ticks)
		/// </summary>
		public FlushInterval FlushInterval
		{
			get { return flushInterval; }
			set { flushInterval = value; }
		}

		/// <summary>
		/// Specifie how the cache content should be returned.
		/// If true a deep copy is returned.
		/// </summary>
		/// <remarks>
		/// Combinaison
		/// IsReadOnly=true/IsSerializable=false : Returned instance of cached object
		/// IsReadOnly=false/IsSerializable=true : Returned coopy of cached object
		/// </remarks>
		public bool IsSerializable
		{
			get { return isSerializable; }
		}

		/// <summary>
		/// Determines if the cache will be used as a read-only cache.
		/// Tells the cache model that is allowed to pass back a reference to the object
		/// existing in the cache.
		/// </summary>
		/// <remarks>
		/// The IsReadOnly properties works in conjonction with the IsSerializable propertie.
		/// </remarks>
		public bool IsReadOnly
		{
			get { return isReadOnly; }
		}
		#endregion

		#region Constructor (s) / Destructor

        /// <summary>
        /// Initializes a new instance of the <see cref="CacheModel"/> class.
        /// </summary>
        /// <param name="name">The name.</param>
        /// <param name="controllerImplementation">The controller implementation.</param>
        /// <param name="properties">The properties.</param>
        public CacheModel(
            string name, 
            string controllerImplementation,
            IDictionary properties)
            : this(name, controllerImplementation, properties, true, false)
        {}

        /// <summary>
        /// Initializes a new instance of the <see cref="CacheModel"/> class.
        /// </summary>
        /// <param name="id">The id.</param>
        /// <param name="implementation">The controller implementation.</param>
        /// <param name="properties">The properties.</param>
        /// <param name="isReadOnly">if set to <c>true</c> [is read only].</param>
        /// <param name="isSerializable">if set to <c>true</c> [is serializable].</param>
        public CacheModel(
            string id,
            string implementation, 
            IDictionary properties,
            bool isReadOnly,
            bool isSerializable) 
		{
            Contract.Require.That(id, Is.Not.Null & Is.Not.Empty).When("retrieving argument id in CacheModel constructor");
            Contract.Require.That(implementation, Is.Not.Null & Is.Not.Empty).When("retrieving argument implementation in CacheModel constructor");
            Contract.Require.That(properties, Is.Not.Null).When("retrieving argument properties in CacheModel constructor");

            this.id = id;
            this.implementation = implementation;
            this.isReadOnly = isReadOnly;
            this.isSerializable = isSerializable;

			lastFlush = DateTime.Now.Ticks;

			try 
			{
				// Build the CacheController
                Type type = TypeUtils.ResolveType(implementation);
				object[] arguments = new object[0];

				controller = (ICacheController)Activator.CreateInstance(type, arguments);
			} 
			catch (Exception e) 
			{
				throw new ConfigurationException("Error instantiating cache controller for cache named '"+id+". Cause: " + e.Message, e);
			}

			//------------ configure Controller---------------------
			try 
			{
				controller.Configure(properties);
			} 
			catch (Exception e) 
			{
				throw new ConfigurationException ("Error configuring controller named '"+id+"'. Cause: " + e.Message, e);
			}
		}


		/// <summary>
		/// Event listener
		/// </summary>
		/// <param name="mappedStatement">A MappedStatement on which we listen ExecuteEventArgs event.</param>
		public void RegisterTriggerStatement(IMappedStatement mappedStatement)
		{
			mappedStatement.Execute +=FlushHandler;
		}
		
		
		/// <summary>
		/// FlushHandler which clear the cache 
		/// </summary>
		/// <param name="sender"></param>
		/// <param name="e"></param>
		private void FlushHandler(object sender, ExecuteEventArgs e)
		{
			if (logger.IsDebugEnabled) 
			{
				logger.Debug("Flush cacheModel named "+id+" for statement '"+e.StatementName+"'");
			}

			Flush();
		}


		/// <summary>
		/// Clears all elements from the cache.
		/// </summary>
		public void Flush() 
		{
			lastFlush = DateTime.Now.Ticks;
			controller.Flush();
		}


		/// <summary>
		/// Adds an item with the specified key and value into cached data.
		/// Gets a cached object with the specified key.
		/// </summary>
		/// <value>The cached object or <c>null</c></value>
		/// <remarks>
		/// A side effect of this method is that is may clear the cache
		/// if it has not been cleared in the flushInterval.
		/// </remarks> 
		public object this [CacheKey key] 
		{
			get
			{
				lock(this) 
				{
					if (lastFlush != NO_FLUSH_INTERVAL
						&& (DateTime.Now.Ticks - lastFlush > flushInterval.Interval)) 
					{
						Flush();
					}
				}

				object value = null;
				lock (GetLock(key)) 
				{
					value = controller[key];
				}

				if(isSerializable && !isReadOnly &&
					(value != NULL_OBJECT && value != null))
				{
					try
					{
						MemoryStream stream = new MemoryStream((byte[]) value);
						stream.Position = 0;
						BinaryFormatter formatter = new BinaryFormatter();
						value = formatter.Deserialize( stream );
					}
					catch(Exception ex)
					{
						throw new IbatisException("Error caching serializable object.  Be sure you're not attempting to use " +
							"a serialized cache for an object that may be taking advantage of lazy loading.  Cause: "+ex.Message, ex);
					}
				}

				lock(syncLock) 
				{
					requests++;
					if (value != null) 
					{
						hits++;
					}
				}

                if (logger.IsDebugEnabled)
                {
                    if (value != null)
                    {
                        logger.Debug(String.Format("Retrieved cached object '{0}' using key '{1}' ", value, key));
                    }
                    else
                    {
                        logger.Debug(String.Format("Cache miss using key '{0}' ", key));
                    }
                }
				return value;
			}
			set
			{
				if (null == value) {value = NULL_OBJECT;}
				if(isSerializable && !isReadOnly && value != NULL_OBJECT)
				{
					try
					{
						MemoryStream stream = new MemoryStream();
						BinaryFormatter formatter = new BinaryFormatter();
						formatter.Serialize(stream, value);
						value = stream.ToArray();
					}
					catch(Exception ex)
					{
						throw new IbatisException("Error caching serializable object. Cause: "+ex.Message, ex);
					}
				}
				controller[key] = value;
                if (logger.IsDebugEnabled)
                {
                    logger.Debug(String.Format("Cache object '{0}' using key '{1}' ", value, key));
                }
			}
		}

		/// <summary>
		/// 
		/// </summary>
		public double HitRatio 
		{
			get 
			{
				if (requests!=0)
				{
					return (double)hits/(double)requests;
				}
				else
				{
					return 0;
				}
			}
		}

		#endregion

		/// <summary>
		/// 
		/// </summary>
		/// <param name="key"></param>
		/// <returns></returns>
		[MethodImpl(MethodImplOptions.Synchronized)]
		public object GetLock(CacheKey key) 
		{
			int controllerId = HashCodeProvider.GetIdentityHashCode(controller);
			int keyHash = key.GetHashCode();
			int lockKey = 29 * controllerId + keyHash;
			object lok =lockMap[lockKey];
			if (lok == null) 
			{
				lok = lockKey; //might as well use the same object
				lockMap[lockKey] = lok;
			}
			return lok;
		}
	}
}
